Add cond-expand from SRFI-0.

bnoordhuis opened the pull request:
Motivation: no cond-expand makes it pretty much impossible to port libraries to Chez Scheme.

I tried to follow the house style as much as possible, apologies if I got it wrong.  Thanks for Chez Scheme, it's amazing!

Closes out #91.

bnoordhuis commented on s/syntax.ss:
```diff
@@ -8931,6 +8931,31 @@
+           (test (cond-expand more ...))
+           (else body ...))]
+      [(_ [name body ...] more ...)
+       (if (eq? 'chez (syntax->datum #'name))
```
This isn't very elegant but I couldn't get it to work by adding `chez` to the syntax-case's literals and adding a pattern, I presume because the symbol doesn't get exported.  If there is a better way to do this, I'll be happy to update the pull request.

dybvig commented:
Thanks for the pull request.  I have some questions and issues to raise before proceeding.

1. Your note says this closes out [#91](https://github.com/cisco/ChezScheme/issues/91).  The requester in that case was looking for R7RS `cond-expand` rather than SRFI-0 `cond-expand`.  Are there any relevant differences?

2. Is it necessary or even sufficient to build `cond-expand` into Chez Scheme?  Presumably, a library or top-level program using`cond-expand` would have to import it from `(chezscheme)` or another non-standard library.  If you have to modify an importing library or top-level program to do that, you might as well modify it to import from an external library that defines a Chez-Scheme-specific `cond-expand`.

3. Does the fact that Chez Scheme looks for libraries first under the extension "chezscheme.sls" already give you what you need?  If nothing else, you can put a Chez-Scheme-specific version of `cond-expand` in "cond-expand.chezscheme.sls" and a generic version in "cond-expand.sls".  Importing libraries can then import `cond-expand` from `(cond-expand)` and remain portable.

4. Informal usage aside, `chez` doesn't make sense on its own.  Please consider an alternative, such as `chez-scheme` or `chezscheme`.

bnoordhuis commented:
Hi, thanks for your comments.

> 1. Your note says this closes out #91. The requester in that case was looking for R7RS cond-expand rather than SRFI-0 cond-expand. Are there any relevant differences?

R7RS stipulates two feature ids that I wasn't sure whether to implement:

1. `r7rs` should evaluate to true only when the runtime is R7RS-compliant.  I don't think this applies to Chez Scheme, in which case evaluating to false is the right thing to do.

2. `(library id)` should evaluate to true when library _id_ is available.  The exact meaning of 'library' seems somewhat underspecified and implementation-specific.  FWIW, I've never seen it used in the wild.

My unscientific sampling suggests that 99% of cond-expand forms are just to detect the Scheme runtime so I figured I'd start out simple.  I can add test cases that explicitly check that the aforementioned features evaluate to false if you want.

> 2. Is it necessary or even sufficient to build cond-expand into Chez Scheme? [...]
> 3. Does the fact that Chez Scheme looks for libraries first under the extension "chezscheme.sls" already give you what you need?

I don't think that would be practical.  Most code I've seen simply assumes that cond-expand is available because that's how it works in other Scheme environments.

I think of it as the Scheme equivalent of `#ifdef __GNUC__`.  Having to jump through additional hoops to make it available would defy its purpose.

The linked lalr-scm pull request is a good example of cond-expand in a single-file library.  Your chezscheme.sls suggestion would destroy the single-file aspect and make porting needlessly complicated.  It would also have to be duplicated in every ported project.

> 4. Informal usage aside, chez doesn't make sense on its own.  Please consider an alternative, such as chez-scheme or chezscheme.

I can change that if you insist but it's similar to how other runtimes announce themselves (e.g., `chibi`, `chicken`, `gauche`, `guile`, `kawa`, etc.)  `chez-scheme` or `chezscheme` would look rather unidiomatic.

dybvig commented:
No identifiers are visible in an R6RS library or top-level program unless they are imported from somewhere.  After looking at lalr-scm, it seems you mean a different notion of library, i.e., unstructured code loaded into the interactive top-level via the `load` procedure.  Instead of supporting this less structured, less reliable, and less efficient packaging mechanism, I'd rather we push toward adoption of R6RS libraries and encourage other implementations to support analogues of the chezscheme.sls extension for implementation-dependent code.

`chezscheme` or `chez-scheme` is more consistent than `chez` for most if not all of the examples you list:  Guile => `guile`, Gauche => `guache`, Kawa => `kawa`, Chez Scheme => `chezscheme`.  I'm not sure about Chibi or Chicken; their web pages refer to them both with and without "Scheme" as part of the name, so `chibi` and `chicken` are as good as `chibischeme` and `chickenscheme`.  Chez Scheme is never properly referred to as "Chez", though people sometimes do so informally.

bnoordhuis commented:
I've renamed the identifier to `chezscheme` and fixed the conflict in LOG.

> Instead of supporting this less structured, less reliable, and less efficient packaging mechanism, I'd rather we push toward adoption of R6RS libraries and encourage other implementations to support analogues of the chezscheme.sls extension for implementation-dependent code.

That seems like an uphill battle if there ever was one.  I think you ought to go for pragmatism over idealism here: having an out-of-the-box cond-expand makes Chez Scheme more accessible to programmers coming from other Schemes.  That's an unequivocally good thing, IMO.

> chezscheme or chez-scheme is more consistent than chez for most if not all of the examples you list: Guile => guile, Gauche => guache, Kawa => kawa, Chez Scheme => chezscheme. 

As mentioned, I've renamed it but I disagree with this line of reasoning: the others are consistently referred to as `<name> Scheme` too, if for no other reason than to disambiguate between the runtime and the word it's named after.  `chezscheme` is simply going to look odd but that's your prerogative.

svenha commented:
Hi all.

Ben, thanks for pushing the very popular SRFI-0 (yes, cond-expand is somewhat ugly, but widespread in many code bases I have seen).
Most Scheme implementations provide a function to set additional feature identifiers. Is this implemented in your pull request, too?

Sven

bnoordhuis commented:
Hi Sven.  No, I implemented just the bare essentials.  I figured I'd start out small for easy reviewing and to make it less contentious.

akeep commented:
Since this is implemented as a single macro, it doesn't seem necessary to include this in the base of Chez Scheme.  It is worth noting that there is a package of Chez Scheme compatible SRFIs in https://github.com/arcfide/chez-srfi, which includes SRFI-0.
